modules/tinymce/src/plugins/link/main/ts/ui/Controls.ts

```diff
@@ -1,57 +1,95 @@
 import { Fun, Optional, Optionals } from '@ephox/katamari';
 
 import Editor from 'tinymce/core/api/Editor';
-import { InlineContent } from 'tinymce/core/api/ui/Ui';
+import { NodeChangeEvent } from 'tinymce/core/api/EventTypes';
+import { InlineContent, Menu, Toolbar } from 'tinymce/core/api/ui/Ui';
 
 import * as Options from '../api/Options';
 import * as Actions from '../core/Actions';
+import * as OpenLink from '../core/OpenLink';
 import * as Utils from '../core/Utils';
 
-const setupButtons = (editor: Editor): void => {
+const openDialog = (editor: Editor) => (): void => {
+  editor.execCommand('mceLink', false, { dialog: true });
+};
+
+const toggleState = (editor: Editor, toggler: (e: NodeChangeEvent) => void): () => void => {
+  editor.on('NodeChange', toggler);
+  return () => editor.off('NodeChange', toggler);
+};
+
+const toggleLinkState = (editor: Editor) => (api: Toolbar.ToolbarToggleButtonInstanceApi | Menu.ToggleMenuItemInstanceApi): () => void => {
+  const updateState = () => {
+    api.setActive(!editor.mode.isReadOnly() && Utils.isInAnchor(editor, editor.selection.getNode()));
+    api.setEnabled(editor.selection.isEditable());
+  };
+  updateState();
+  return toggleState(editor, updateState);
+};
+
+const toggleLinkMenuState = (editor: Editor) => (api: Menu.MenuItemInstanceApi): () => void => {
+  const updateState = () => {
+    api.setEnabled(editor.selection.isEditable());
+  };
+  updateState();
+  return toggleState(editor, updateState);
+};
+
+const toggleRequiresLinkState = (editor: Editor) => (api: Toolbar.ToolbarButtonInstanceApi | Menu.MenuItemInstanceApi): () => void => {
+  const hasLinks = (parents: Node[]) => Utils.hasLinks(parents) || Utils.hasLinksInSelection(editor.selection.getRng());
+  const parents = editor.dom.getParents(editor.selection.getStart());
+  const updateEnabled = (parents: Node[]) => {
+    api.setEnabled(hasLinks(parents) && editor.selection.isEditable());
+  };
+  updateEnabled(parents);
+  return toggleState(editor, (e) => updateEnabled(e.parents));
+};
+
+const setupButtons = (editor: Editor, openLink: OpenLink.LinkSelection): void => {
   editor.ui.registry.addToggleButton('link', {
     icon: 'link',
     tooltip: 'Insert/edit link',
-    onAction: Actions.openDialog(editor),
-    onSetup: Actions.toggleLinkState(editor),
-    shortcut: 'Meta+K'
+    shortcut: 'Meta+K',
+    onAction: openDialog(editor),
+    onSetup: toggleLinkState(editor)
   });
 
   editor.ui.registry.addButton('openlink', {
     icon: 'new-tab',
     tooltip: 'Open link',
-    onAction: Actions.gotoSelectedLink(editor),
-    onSetup: Actions.toggleGotoLinkState(editor)
+    onAction: openLink.gotoSelectedLink,
```

Comment:
Yes since its the same menu item as in the context menu, and the toolbar buttons follow the same onActions. If you've just right clicked a single link, opening the context menu, then click open link from the toolbar, it'll open the right-clicked link. I've just pushed a commit to clear it on SelectionChange, so it will always fallback to checking the current selection when the context menu is closed.

Reply:
Resolving this conversation as it's been addressed. I had concerns that `selectedLink` wasn't set when the context menu wasn't opened, and clicking on the toolbar button would do nothing since the `selectedLink` is Optional.none.